main.go

```diff
@@ -209,6 +247,162 @@ func GetDfoActiveContactList(apiUrl string) []byte {
 	return responseData
 }
 
+// GetDfoAuthToken calls the DFO api POST engager/2.0/token to get a bearer token for subsequent DFO api calls
+func GetDfoAuthToken(apiUrl string) DfoAuthTokenObj {
+	contentType := "application/json"
+	var dfoAuthTokenBody DfoAuthTokenBody
+	dfoAuthTokenBody.GrantType = "client_credentials"
+
+	clientId := GetTenantData(clientIdRequest)
+
+	//Validate a valid ClientId was input (just checks string length)
+	validClientId := false
+	for !validClientId {
+		validClientId = Validate(clientId, false, validClientId)
+		if validClientId {
+			dfoAuthTokenBody.ClientId = clientId
+			break
+		} else {
+			clientId = GetTenantData(clientIdRequest)
+		}
+	}
+
+	clientSecret := GetTenantData(clientSecretRequest)
+
+	//Validate a valid ClientSecret was input (just checks string length)
+	validClientSecret := false
+	for !validClientSecret {
+		validClientSecret = Validate(clientSecret, false, validClientSecret)
+		if validClientSecret {
+			dfoAuthTokenBody.ClientSecret = clientSecret
+			break
+		} else {
+			clientSecret = GetTenantData(clientSecretRequest)
+		}
+	}
+
+	bodyJson, _ := json.Marshal(dfoAuthTokenBody)
+	reader := bytes.NewReader(bodyJson)
+	var responseData []byte
+
+	response, err := http.Post(apiUrl, contentType, reader)
+
+	if err != nil {
+		fmt.Print(err.Error())
+		os.Exit(1)
+	}
+
+	if response.StatusCode == 200 {
+		responseData, err = ioutil.ReadAll(response.Body)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	var dfoAuthTokenObj DfoAuthTokenObj
+	if responseData != nil {
+		err := json.Unmarshal(responseData, &dfoAuthTokenObj)
+		if err != nil {
+			fmt.Println("Cannot unmarshal dfoAuthToken")
+		}
+	} else {
+		fmt.Println("DfoAuthToken was null or empty.")
+	}
+	return dfoAuthTokenObj
+}
+
+// GetDeltaList loops through the known contacts from DMW and compares them to the active contacts in DFO and creates a list of contacts that need to be updated
+func GetDeltaList(dmwKnownContacts DmwKnownContacts, dfoActiveContacts DfoActiveContacts) []byte {
+	found := false
+	var newState int32
+	var updateDateStr string
+	type DmwContacts []DmwKnownContact
+	deltaArray := DmwContacts{}
+	var deltaList DmwKnownContacts
+
+	// Loop through DmwKnownContacts and check each contact data in DfoActiveContacts to see if we find a match
+	for _, contact := range dmwKnownContacts.Contacts {
+		for _, data := range dfoActiveContacts.Data {
```

Comment:
This is n*n algorithm, so not the most optimal. If you're never dealing with a lot of contacts, this is probably ok, but it could cause problems as the number grows. One option is to sort both lists (n*log n each), then iterate through both "simultaneously" since everything is in order (2n).

Reply:
Looking at Big O.  Professor Owens would be so proud of you!

Reply:
You keep track of two indexes and walk both sorted lists simultaneously. Let's say the indexes are i and j. If DMW[i] == DFO[j] then there's a match and you can increment both indexes. If DMW[i] < DFO[j], then DMW has a contact that isn't in DFO; increment only i. If DMW[i] > DFO[j] then DFO has a contact that isn't in DMW; increment only j.